README.md
# NAME

MooseX::AttributeFilter - MooX::AttributeFilter with cute antlers OwO

# SYNOPSIS

    package My::Class;
    use Moose;
    use MooseX::AttributeFilter;
    
    has field => (
        is     => 'rw',
        filter => 'filterField',
    );
    
    sub filterField {
        my $this = shift;
        return "filtered($_[0])";
    }
    
    package main;
    my $obj = My::Class->new( field => "initial" );
    $obj->field eq "filtered(initial)"; # True!

# DESCRIPTION

MooseX::AttributeFilter is a port of [MooX::AttributeFilter](https://metacpan.org/pod/MooX::AttributeFilter) to Moose.

Some parts don't work correctly in mutable classes.
Mutable classes are slow anyway.

# LICENSE

Copyright (C) Little Princess Kitten <3.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

KITTEN

[https://metacpan.org/author/KITTEN](https://metacpan.org/author/KITTEN)

[https://github.com/icklekitten](https://github.com/icklekitten)


Changes
Revision history for Perl extension MooseX-AttributeFilter

0.07

    - more tolerant looking for filter method by name

0.06

    - fix NAME for MooseX::AttributeFilter::Trait::Attribute::Role docs

0.05

    - better docs

0.04

    - tests for filtered attributes in roles
    - tests for meta
    - documentation for meta
    - ily

0.03

    - bump required Moose version even more D:

0.02

    - bump required Moose version

0.01

    - original version


lib/MooseX/AttributeFilter/Trait/Attribute.pm
package MooseX::AttributeFilter::Trait::Attribute;
use 5.008009;
use strict;
use warnings;

our $VERSION = "0.07";

use Moose::Role;

has filter => (
    is  => 'ro',
    isa => 'CodeRef|Str',
    predicate => 'has_filter',
);

has bypass_filter_method_check => (
    is  => 'ro',
    isa => 'Bool',
    default => 0,
);

after _process_options => sub {
    my $this = shift;
    my ($name, $options) = @_;
    
    return unless defined $options->{filter};
    
    if ($options->{filter} eq '1') {
        $options->{filter} = "_filter_${name}";
    }
};

before install_accessors => sub {
    my $this = shift;
    return if $this->bypass_filter_method_check;
    my $filter = $this->filter;
    if (defined $filter and not ref $filter) {
        my $class  = $this->associated_class;
        my $method = $class->find_method_by_name($filter)
            || $class->name->can($filter)
            || eval {
                no strict 'refs';
                exists &{ $class->name . "::" . $filter }
                    ? \&{ $class->name . "::" . $filter }
                    : undef;
            };
        
        die sprintf(
            "No filter method '%s' defined for %s attribute '%s'",
            $filter,
            $class->name,
            $this->name,
            ) if !$method;
    }
};

around _coerce_and_verify => sub {
    my $next = shift;
    my $this = shift;
    my ($value, $instance) = @_;
    my $filter = $this->filter;
    if ($filter) {
        
        # Need to figure out if we're being called by non-immutable
        # constructor. This is ugly, but searching the call stack works.
        my $in_ctor = 0;
        for my $level (0 .. 8) {  # should be enough
            my (undef, undef, undef, $sub) = caller($level);
            if ($sub eq 'Moose::Meta::Attribute::initialize_instance_slot') {
                $in_ctor = 1;
                last;
            }
        }
        
        $value = $instance->$filter($value, $in_ctor ? () : $this->get_value($instance));
    }
    $this->$next($value, $instance);
};

has _filter_ix => (
    is  => 'rw',
    isa => 'Int',
);

sub _inline_filter {
    my $this = shift;
    my ($instance, $value, $filtered, $for_constructor) = @_;
    
    my @code;
    my $filter = $this->filter;
    if (ref $filter) {
        if (not defined $this->_filter_ix) {
            # going via metaobject would be smarter
            # but this is faster. xD
            push(our @FILTERS, $filter);
            $this->_filter_ix($#FILTERS);
        }
        push @code, sprintf(
            'my $filter = $%s::FILTERS[%d];',
            __PACKAGE__,
            $this->_filter_ix,
        );
        push @code, sprintf(
            $for_constructor
                ? 'my %s = %s->%s(%s);'
                : 'my %s = %s->%s(%s, %s);',
            $filtered,
            $instance,
            '$filter',
            $value,
            $for_constructor
                ? ()
                : $this->_inline_instance_get($instance),
        ),
    }
    elsif (defined $filter) {
        push @code, sprintf(
            $for_constructor
                ? 'my %s = %s->%s(%s);'
                : 'my %s = %s->%s(%s, %s);',
            $filtered,
            $instance,
            $filter,
            $value,
            $for_constructor
                ? ()
                : $this->_inline_instance_get($instance),
        ),
    }
    
    return @code;
}

around _inline_set_value => sub {
    my $next = shift;
    my $this = shift;
    my ($instance, $value, $tc, $coercion, $message, $for_constructor) = @_;    
    return $this->$next(@_) unless $this->has_filter;
    my @code = (
        $this->_inline_filter($instance, $value, '$filtered', $for_constructor),
        $this->$next($instance, '$filtered', $tc, $coercion, $message, $for_constructor),
    );
    #use Data::Dumper;
    #warn Dumper \@code;
    return @code;
};

around _inline_init_from_default => sub {
    my $next = shift;
    my $this = shift;
    my ($instance, $default, $tc, $coercion, $message, $for_lazy) = @_;
    my $filtered = '$filtered';
    return $this->$next(@_) unless $this->has_filter;
    
    my @code = (
        $this->_inline_generate_default($instance, $default),
        $this->_inline_filter($instance, $default, $filtered, $for_lazy),
        $this->has_type_constraint
            ? ($this->_inline_check_coercion($filtered, $tc, $coercion, $for_lazy),
               $this->_inline_check_constraint($filtered, $tc, $message, $for_lazy))
            : (),
        $this->_inline_init_slot($instance, $filtered),
        $this->_inline_weaken_value($instance, $filtered),
    );
    #use Data::Dumper;
    #warn Dumper \@code;
    return @code;
};

1;
__END__

=encoding utf-8

=head1 NAME

MooseX::AttributeFilter::Trait::Attribute - trait for filtered attributes

=head1 SYNOPSIS

    package My::Class;
    use Moose;
    use MooseX::AttributeFilter;
    
    has field => (
        is     => 'rw',
        filter => 'filterField',
    );
    
    sub filterField {
        my $this = shift;
        return "filtered($_[0])";
    }
    
    package main;
    My::Class->meta->get_attribute("field")->has_filter;  # true

=head1 DESCRIPTION

MooseX::AttributeFilter::Trait::Attribute is a trait for L<Moose::Meta::Attribute>.
L<MooseX::AttributeFilter> automatically applies it to all attributes, but it
acts as no-op if attribute does not use C<filter> option.

=head2 Methods

=over

=item C<filter>

Returns the value of the C<filter> option. This may be a string (method name)
or coderef or undef.

=item C<has_filter>

Boolean.

=item C<bypass_filter_method_check>

Boolean.

    has attr => (
        is      => 'rw',
        filter  => 'my_filter_method',
        bypass_filter_method_check => 1,
    );

Can be used to make MooseX::AttributeFilter::Trait::Attribute not throw
error message in case of C<my_filter_method> method not exists. (Will still
throw error if is called and cannot find method.) Used for edge cases like
when filter method will be provided by C<AUTOLOAD> or later added by trait.

=back

=head1 SEE ALSO

L<MooseX::AttributeFilter>.

=head1 LICENSE

Copyright (C) 2018 Little Princess Kitten <kitten@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

KITTEN <kitten@cpan.org>

L<https://metacpan.org/author/KITTEN>

L<https://github.com/icklekitten>

<3

=cut



lib/MooseX/AttributeFilter/Trait/Attribute/Role.pm
package MooseX::AttributeFilter::Trait::Attribute::Role;
use 5.008009;
use strict;
use warnings;

our $VERSION = "0.07";

use Moose::Role;

has filter => (
    is  => 'ro',
    isa => 'CodeRef|Str',
    predicate => 'has_filter',
);

has bypass_filter_method_check => (
    is  => 'ro',
    isa => 'Bool',
    default => 0,
);

1;

__END__

=encoding utf-8

=head1 NAME

MooseX::AttributeFilter::Trait::Attribute::Role - trait for filtered attributes in roles

=head1 SYNOPSIS

    package My::Role;
    use Moose::Role;
    use MooseX::AttributeFilter;
    
    has field => (
        is     => 'rw',
        filter => 'filterField',
    );
    
    sub filterField {
        my $this = shift;
        return "filtered($_[0])";
    }
    
    package My::Class;
    use Moose;
    with 'My::Role';
    
    package main;
    My::Role->meta->get_attribute("field")->has_filter;  # true

=head1 DESCRIPTION

This basically does nothing but is used when a role containing filtered
attributes is composed into a class. 

=head2 Methods

It has some things for introspection tho. c:

=over

=item C<filter>

Returns the value of the C<filter> option. This may be a string (method name)
or coderef or undef.

=item C<has_filter>

Boolean.

=item C<bypass_filter_method_check>

Boolean.

=back

=head1 SEE ALSO

L<MooseX::AttributeFilter>.

=head1 LICENSE

Copyright (C) 2018 Little Princess Kitten <kitten@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

KITTEN <kitten@cpan.org>

L<https://metacpan.org/author/KITTEN>

L<https://github.com/icklekitten>

<3

=cut



lib/MooseX/AttributeFilter.pm
package MooseX::AttributeFilter;
use 5.008009;
use strict;
use warnings;

our $VERSION = "0.07";

use Moose ();
use Moose::Exporter;

Moose::Exporter->setup_import_methods();

sub init_meta {
    my ($class_name, %args) = @_;
    my $caller = $args{for_class};
    my $trait  = 'MooseX::AttributeFilter::Trait::Attribute';
    
    Moose::Util::MetaRole::apply_metaroles(
        for             => $caller,
        class_metaroles => {
            attribute         => [ $trait ],
        },
        role_metaroles => {
            attribute         => [ $trait . '::Role' ],
            applied_attribute => [ $trait ],
        },
        parameter_metaroles => {
            applied_attribute => [ $trait ],
        },
        parameterized_role_metaroles => {
            applied_attribute => [ $trait ],
        },
    );
}

1;
__END__

=encoding utf-8

=head1 NAME

MooseX::AttributeFilter - MooX::AttributeFilter with cute antlers OwO

=head1 SYNOPSIS

    package My::Class;
    use Moose;
    use MooseX::AttributeFilter;
    
    has field => (
        is     => 'rw',
        filter => 'filterField',
    );
    
    sub filterField {
        my $this = shift;
        return "filtered($_[0])";
    }
    
    package main;
    my $obj = My::Class->new( field => "initial" );
    $obj->field eq "filtered(initial)"; # True!

=head1 DESCRIPTION

MooseX::AttributeFilter is a port of L<MooX::AttributeFilter> to L<Moose>.

Filter is like a C<coerce> sub but is called as a method so can see object instance.

Filter is like a C<trigger> but is called before attribute value is set.

=head1 BUGS

Some parts don't work correctly in mutable classes. Mutable classes are slow anyway.

L<https://rt.cpan.org/Dist/Display.html?Queue=MooseX-AttributeFilter>

=head1 CUTE

=begin html

<img height="680" width="500" src="https://data.whicdn.com/images/129435330/large.jpg" alt="cute kitty girl" />

=end html

Cute.

=head1 SEE ALSO

L<MooX::AttributeFilter>, L<Moose>.

L<MooseX::AttributeFilter::Trait::Attribute>,
L<MooseX::AttributeFilter::Trait::Attribute::Role>.

=head1 LICENSE

Copyright (C) 2018 Little Princess Kitten <kitten@cpan.org>

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

KITTEN <kitten@cpan.org>

L<https://metacpan.org/author/KITTEN>

L<https://github.com/icklekitten>

<3

=cut

